First pass

pengale opened the pull request:
This is pretty close to done. I'd like opinions on which of the remaining TODOs are actually blockers. Thank you.

@juju-solutions/bigdata

pengale committed:
Brought over LICENSE, icon.svg, and metadata.yaml from existing charm.

Stubbed out files that are going to change more for this bigtop-based charm.

pengale committed:
Got Zookeeper charm basically "working".

"Working" in this case means that the charm executes, and Bigtop seems
to do its thing. A lot of methods are still missing, though, and we
don't actually have a *useful* Zookeeper service just yet. Stay tuned ...

pengale committed:
Added code for adjusting quorum.

Ported over quorum related methods from old charm, and adjusted for
Bigtop.

pengale committed:
Ported README over from old charm.

Needs some updating, but appears to be mainly still relevant.

pengale committed:
Ported tests over from old charm -- it looks like they are still valid for the new charm.

pengale committed:
Got rid of RESTful interface TODOs, plus cleanup.

pengale committed:
Semi-fixed quorum stuff.

Discovered a bug in the quorum functions I wrote. Partly fixed, though
the server status function isn't working properly. Need to troubleshoot
and figure out what's wrong with the way that we are writing the config.

pengale commented on reactive/zookeeper.py:
```diff
@@ -0,0 +1,60 @@
+    zk = Zookeeper()
+    zk.increase_quorum(nodes)
+    #restart_zookeeper_if_config_changed()  # Not necessary due to puppet
+    zkpeer.dismiss_joined()
```
This _almost_ works. I get what looks like the right stuff written out the the zoo.cfg, and the server restarts.

However, `/usr/lib/zookeeper/bin/zkServer.sh status` doesn't work. It claims to have an error contacting the service, but the service is running.

I bet that it's something subtlely wrong with the way that I'm writing the config file. Haven't figured out what it is yet, though :-/

ktsakalozos replied:
Indeed, the process is up but the Zookeeper instances do not form a quorum.

If you try to connect to zk with zkCli.sh you fail with:
`
2016-06-02 11:25:32,733 [myid:] - INFO  [main-SendThread(localhost:2181):ClientCnxn$SendThread@852] - Socket connection established to localhost/127.0.0.1:2181, initiating session
2016-06-02 11:25:32,738 [myid:] - INFO  [main-SendThread(localhost:2181):ClientCnxn$SendThread@1098] - Unable to read additional data from server sessionid 0x0, likely server has closed socket, closing socket connection and attempting reconnect
[zk: localhost:2181(CONNECTING) 0] 2016-06-02 11:25:34,803 [myid:] - INFO  [main-SendThread(localhost:2181):ClientCnxn$SendThread@975] - Opening socket connection to server localhost/127.0.0.1:2181. Will not attempt to authenticate using SASL (unknown error)
`

In /var/log/zookeeper/zookeeper.log we get this exception:
`
2016-06-02 11:31:30,338 [myid:0] - WARN  [RecvWorker:0:QuorumCnxManager$RecvWorker@780] - Connection broken for id 0, my id = 0, error = 
java.io.EOFException
        at java.io.DataInputStream.readInt(DataInputStream.java:392)
        at org.apache.zookeeper.server.quorum.QuorumCnxManager$RecvWorker.run(QuorumCnxManager.java:765)
2016-06-02 11:31:30,338 [myid:0] - WARN  [RecvWorker:0:QuorumCnxManager$RecvWorker@783] - Interrupting SendWorker
2016-06-02 11:31:30,339 [myid:0] - WARN  [SendWorker:0:QuorumCnxManager$SendWorker@697] - Interrupted while waiting for message on queue
java.lang.InterruptedException
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.reportInterruptAfterWait(AbstractQueuedSynchronizer.java:2014)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2088)
        at java.util.concurrent.ArrayBlockingQueue.poll(ArrayBlockingQueue.java:418)
        at org.apache.zookeeper.server.quorum.QuorumCnxManager.pollSendQueue(QuorumCnxManager.java:849)
        at org.apache.zookeeper.server.quorum.QuorumCnxManager.access$500(QuorumCnxManager.java:64)
        at org.apache.zookeeper.server.quorum.QuorumCnxManager$SendWorker.run(QuorumCnxManager.java:685)
`

There are a couple of things you should investigate including if we are affected by this: https://issues.apache.org/jira/browse/ZOOKEEPER-2106

For sure this is a blocker :(

andrewdmcleod commented on layer.yaml:
```diff
@@ -0,0 +1,21 @@
+includes:
+  - 'layer:basic'
```
minor - basic is already included in apache-bigtop-base

pengale committed:
Added consistent sorting for Zookeeper nodes.

This addresses an error message about Zookeeper not joing quorum do to
"smaller ids".

We're now getting "connection refused" errors, but at least we're making
progress ...

pengale committed:
Slightly less broken way of specifying peer ips.

This still doesn't work, but I think that it's closer.

I pinged the Bigtop mailing list for help.

pengale committed:
Final fixes for quorum.

This requires a corresponding patch in the base layer.

ktsakalozos commented:
This layer has to be moved as a branch under juju-solutions/bigtop so that it gets merged with the rest of the layers.

Here is where we put the layers we buid:
https://github.com/juju-solutions/bigtop/tree/master/bigtop-packages/src/charm
Have a look at the ignite example https://github.com/juju-solutions/bigtop/commit/aa93f8c8b1effafd730020e551e26a135ee6d83f

Looking at the list of mudules this layer should be called hadoop-zookeeper: https://github.com/juju-solutions/bigtop/tree/master/bigtop-deploy/puppet/modules

ktsakalozos commented on lib/charms/layer/zookeeper.py:
```diff
@@ -0,0 +1,158 @@
+
+    '''
+    return (unit.split("/")[1], "{ip}:2888:3888".format(ip=node_ip))
+
```
There are a couple of lint errors. Here is how you can fix them. As soon as you have done a "charm build" you go to the output directory (something like ../../trusty/apache-bigtop-zookeeper) and you do a "make lint" the make rule comes from the basic layer (its magic :) )

ktsakalozos commented on reactive/zookeeper.py:
```diff
@@ -0,0 +1,51 @@
+    puppet will start the service, as a side effect.
+
+    '''
+    zookeeper = Zookeeper()
```
At the very beginning of a reactive method we would want to inform the use/admin what is going on. So you could have a call similar to this: "hookenv.status_set('waiting', 'Installing Zookeeper')"

ktsakalozos commented on reactive/zookeeper.py:
```diff
@@ -0,0 +1,51 @@
+    Add the unit that just joined, restart Zookeeper, and remove the
+    '.joined' state so we don't fall in here again (until another peer joins).
+    """
+    nodes = zkpeer.get_nodes()  # single node since we dismiss .joined below
```
Same as above. Only no you would need to do:

"hookenv.status_set('waiting', 'Configuring Zookeeper')"
Then you would do the install calls that take a considerable amount of time (calls that include puppet apply)
And at the end you set the state to Ready again:
"hookenv.status_set('active', 'Ready')"

ktsakalozos commented on metadata.yaml:
```diff
@@ -0,0 +1,17 @@
+name: apache-bigtop-zookeeper
```
The name of the layer needs to be the name of the bigtop component, "hadoop-zookeeper".

ktsakalozos commented on tests/01-basic-deployment.py:
```diff
@@ -0,0 +1,24 @@
+
+    def test_deploy(self):
+        self.d = amulet.Deployment(series='trusty')
+        self.d.add('apache-zookeeper', 'apache-zookeeper')
```
The second parameter here needs to be the layer name as defined in the metadata.yamle, or else we will be testing another layer

ktsakalozos commented on tests/10-deploy:
```diff
@@ -0,0 +1,28 @@
+    def setUpClass(cls):
+        cls.d = amulet.Deployment(series='trusty')
+
+        cls.d.add('zookeeper', charm='cs:trusty/apache-zookeeper', units=3)
```
Same as above

ktsakalozos commented on reactive/zookeeper.py:
```diff
@@ -0,0 +1,51 @@
+    zkpeer.dismiss_departed()
+
+# TODO: add zookeeper REST
+# TODO: after rest, add client functions from old charm
```
We can live without REST but we need the client functions of the old charm that hook the zookeeper interface. Or else this charm is not usable.

pengale committed:
Addressed comments in PR.

Fixed charm name. Also fixed lint errors, added status messages, and
ported over client method from old charm.
